driver/chain_syncer.go

```diff
@@ -0,0 +1,120 @@
+package driver
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/taikoxyz/taiko-client/metrics"
+	eventIterator "github.com/taikoxyz/taiko-client/pkg/chain_iterator/event_iterator"
+	"github.com/taikoxyz/taiko-client/pkg/rpc"
+	txListValidator "github.com/taikoxyz/taiko-client/pkg/tx_list_validator"
+)
+
+type L2ChainSyncer struct {
+	ctx                           context.Context
+	state                         *State                           // Driver's state
+	rpc                           *rpc.Client                      // L1/L2 RPC clients
+	throwawayBlocksBuilderPrivKey *ecdsa.PrivateKey                // Private key of L2 throwaway blocks builder
+	txListValidator               *txListValidator.TxListValidator // Transactions list validator
+	// Try P2P beacon-sync if current node is behind of  the protocol's latest verified block head
+	p2pSyncVerifiedBlocks       bool
+	lastSyncedVerifiedBlockHash common.Hash
+	lastSyncedVerifiedBlockID   *big.Int
+	beaconSyncTriggered         bool
+}
+
+// NewL2ChainSyncer creates a new chain syncer instance.
+func NewL2ChainSyncer(
+	ctx context.Context,
+	rpc *rpc.Client,
+	state *State,
+	throwawayBlocksBuilderPrivKey *ecdsa.PrivateKey,
+	p2pSyncVerifiedBlocks bool,
+) (*L2ChainSyncer, error) {
+	return &L2ChainSyncer{
+		ctx:                           ctx,
+		rpc:                           rpc,
+		state:                         state,
+		throwawayBlocksBuilderPrivKey: throwawayBlocksBuilderPrivKey,
+		txListValidator: txListValidator.NewTxListValidator(
+			state.maxBlocksGasLimit.Uint64(),
+			state.maxBlockNumTxs.Uint64(),
+			state.maxTxlistBytes.Uint64(),
+			state.minTxGasLimit.Uint64(),
+			rpc.L2ChainID,
+		),
+		p2pSyncVerifiedBlocks: p2pSyncVerifiedBlocks,
+	}, nil
+}
+
+// Sync performs a sync operation to L2 node's local chain.
+func (s *L2ChainSyncer) Sync(l1End *types.Header) error {
+	// If current L2 node's chain is behind of the protocol's latest verified block head, and the
+	// `P2PSyncVerifiedBlocks` flag is set, try triggering a beacon-sync in L2 node to catch up the
+	// latest verified block head.
+	// TODO: check whether the engine is not syncing through P2P (eth_syncing)
+	if s.p2pSyncVerifiedBlocks && s.state.getLastVerifiedBlock().Height.Uint64() > 0 && !s.AheadOfProtocolVerifiedHead() {
+		if err := s.TriggerBeaconSync(); err != nil {
+			return fmt.Errorf("trigger beacon-sync error: %w", err)
+		}
+
+		return nil
+	}
+
+	if s.beaconSyncTriggered {
+		log.Info("Swtich to insert pending blocks one by one")
+
+		l2Head, err := s.rpc.L2.HeaderByNumber(s.ctx, nil)
```

Comment:
l2engine.ForkchoiceStateV1 is an asynchronous call, L2 p2p synchronization has not been completed here, it may need to wait here